File: goopylib/maths/easing.py
"""
Module defining common easing functions
"""

from typing import Callable
import goopylib.ext.easing as __easing


class EasingType:
    """Generic Easing Type"""

    def __call__(self, t: float) -> float:
        """
        Raises:
            TypeError: easing argument must be a float
        """

        raise NotImplementedError()


def ease_linear() -> EasingType:
    r"""Returns a linear ease

    .. math::

        \text{ease\_linear}(t) = t

    Returns:
        a linear ease callable
    """
    return __easing.ease_linear()


def ease_poly(power: float = 6) -> EasingType:
    r"""Returns a polynomial ease

    .. math::

        \text{ease\_poly}(t) = t^{\text{power}}

    Args:
        power (float): The exponent defining the easing.

    Returns:
        a polynomial in-out ease callable
    """
    return __easing.ease_poly(power)


def ease_poly_in(power: float = 6) -> EasingType:
    r"""Returns a polynomial in ease

    .. math::

        \text{ease\_poly}(t) = t^{\text{power}}

    Args:
        power (float): The exponent defining the easing.

    Returns:
        a polynomial in ease callable
    """
    return __easing.ease_poly_in(power)


def ease_poly_out(power: float = 6) -> EasingType:
    r"""Returns a polynomial out ease

    .. math::

        \text{ease\_poly}(t) = t^{\text{power}}

    Args:
        power (float): The exponent defining the easing.

    Returns:
        a polynomial out ease callable
    """
    return __easing.ease_poly_out(power)


def ease_quad() -> EasingType:
    r"""Returns a quadratic ease

    .. math::

        \text{ease\_quad}(t) = t^{2}

    Returns:
        a quadratic in-out ease callable
    """
    return __easing.ease_quad()


def ease_quad_in() -> EasingType:
    r"""Returns a quadratic in ease

    .. math::

        \text{ease\_quad}(t) = t^{2}

    Returns:
        a quadratic in ease callable
    """
    return __easing.ease_quad_in()


def ease_quad_out() -> EasingType:
    r"""Returns a quadratic out ease

    .. math::

        \text{ease\_quad}(t) = t^{2}

    Returns:
        a quadratic out ease callable
    """
    return __easing.ease_quad_out()


def ease_cubic() -> EasingType:
    r"""Returns a cubic ease

    .. math::

        \text{ease\_cubic}(t) = t^{3}

    Returns:
        a cubic in-out ease callable
    """
    return __easing.ease_cubic()


def ease_cubic_in() -> EasingType:
    r"""Returns a cubic in ease

    .. math::

        \text{ease\_cubic}(t) = t^{3}

    Returns:
        a cubic in ease callable
    """
    return __easing.ease_cubic_in()


def ease_cubic_out() -> EasingType:
    r"""Returns a cubic out ease

    .. math::

        \text{ease\_cubic}(t) = t^{3}

    Returns:
        a cubic out ease callable
    """
    return __easing.ease_cubic_out()


def ease_quart() -> EasingType:
    r"""Returns a quartic ease

    .. math::

        \text{ease\_quart}(t) = t^{4}

    Returns:
        a quartic in-out ease callable
    """
    return __easing.ease_quart()


def ease_quart_in() -> EasingType:
    r"""Returns a quartic in ease

    .. math::

        \text{ease\_quart}(t) = t^{4}

    Returns:
        a quartic in ease callable
    """
    return __easing.ease_quart_in()


def ease_quart_out() -> EasingType:
    r"""Returns a quartic out ease

    .. math::

        \text{ease\_quart}(t) = t^{4}

    Returns:
        a quartic out ease callable
    """
    return __easing.ease_quart_out()


def ease_quint() -> EasingType:
    r"""Returns a quintic ease

    .. math::

        \text{ease\_quint}(t) = t^{5}

    Returns:
        a quintic in-out ease callable
    """
    return __easing.ease_quint()


def ease_quint_in() -> EasingType:
    r"""Returns a quintic in ease

    .. math::

        \text{ease\_quint}(t) = t^{5}

    Returns:
        a quintic in ease callable
    """
    return __easing.ease_quint_in()


def ease_quint_out() -> EasingType:
    r"""Returns a quintic out ease

    .. math::

        \text{ease\_quint}(t) = t^{5}

    Returns:
        a quintic out ease callable
    """
    return __easing.ease_quint_out()


def ease_circle() -> EasingType:
    r"""Returns a circular ease

    .. math::

        \text{ease\_circle}(t) = 1 - \sqrt{1 - t^2}

    Returns:
        an circular in-out ease callable
    """
    return __easing.ease_circle()


def ease_circle_in() -> EasingType:
    r"""Returns a circular in ease

    .. math::

        \text{ease\_circle}(t) = 1 - \sqrt{1 - t^2}

    Returns:
        an circular in ease callable
    """
    return __easing.ease_circle_in()


def ease_circle_out() -> EasingType:
    r"""Returns a circular out ease

    .. math::

        \text{ease\_circle}(t) = 1 - \sqrt{1 - t^2}

    Returns:
        an circular out ease callable
    """
    return __easing.ease_circle_out()


def ease_sin() -> EasingType:
    r"""Returns a sine ease

    .. math::

        \text{ease\_sin}(t) = 1 - \text{cos}(t\pi/2)

    Returns:
        a sine in-out ease callable
    """
    return __easing.ease_sin()


def ease_sin_in() -> EasingType:
    r"""Returns a sine in ease

    .. math::

        \text{ease\_sin}(t) = 1 - \text{cos}(t\pi/2)

    Returns:
        a sine in ease callable
    """
    return __easing.ease_sin_in()


def ease_sin_out() -> EasingType:
    r"""Returns a sine out ease

    .. math::

        \text{ease\_sin}(t) = 1 - \text{cos}(t\pi/2)

    Returns:
        a sine out ease callable
    """
    return __easing.ease_sin_out()


def ease_exp(base: float = 2.718281828459045) -> EasingType:
    r"""Returns an exponential ease

    .. math::

        \text{ease\_exp}(t) = \text{base}^{10(t - 1)}

    Args:
        base (float): The base of the exponential function.

    Returns:
        an exponential in-out ease callable
    """
    return __easing.ease_exp(base)


def ease_exp_in(base: float = 2.718281828459045) -> EasingType:
    r"""Returns an exponential in ease

    .. math::

        \text{ease\_exp}(t) = \text{base}^{10(t - 1)}

    Args:
        base (float): The base of the exponential function.

    Returns:
        an exponential in ease callable
    """
    return __easing.ease_exp_in(base)


def ease_exp_out(base: float = 2.718281828459045) -> EasingType:
    r"""Returns an exponential out ease

    .. math::

        \text{ease\_exp}(t) = \text{base}^{10(t - 1)}

    Args:
        base (float): The base of the exponential function.

    Returns:
        an exponential out ease callable
    """
    return __easing.ease_exp_out(base)


def ease_back(factor: float = 1.70158) -> EasingType:
    r"""Returns a back ease

    .. math::

        \text{ease\_back}(t) = (\text{factor} + 1)\times t^3 - \text{factor}\times t^2

    Args:
        factor (float): The scale of the easing's stretch.

    Returns:
        a back in-out ease callable
    """
    return __easing.ease_back(factor)


def ease_back_in(factor: float = 1.70158) -> EasingType:
    r"""Returns a back in ease

    .. math::

        \text{ease\_back}(t) = (\text{factor} + 1)\times t^3 - \text{factor}\times t^2

    Args:
        factor (float): The scale of the easing's stretch.

    Returns:
        a back in ease callable
    """
    return __easing.ease_back_in(factor)


def ease_back_out(factor: float = 1.70158) -> EasingType:
    r"""Returns a back out ease

    .. math::

        \text{ease\_back}(t) = (\text{factor} + 1)\times t^3 - \text{factor}\times t^2

    Args:
        factor (float): The scale of the easing's stretch.

    Returns:
        a back out ease callable
    """
    return __easing.ease_back_out(factor)


def ease_elastic(factor: int = 3) -> EasingType:
    r"""Returns an elastic ease

    .. math::

        \text{ease\_back}(t) = 2^{10(t - 1)}\text{sin}(t\pi/2\times(1 + 4\times\text{factor)})

    Args:
        factor (int): The number of elastic oscillations.

    Returns:
        an elastic in-out ease callable
    """
    return __easing.ease_elastic(factor)


def ease_elastic_in(factor: int = 3) -> EasingType:
    r"""Returns an elastic in ease

    .. math::

        \text{ease\_back}(t) = 2^{10(t - 1)}\text{sin}(t\pi/2\times(1 + 4\times\text{factor)})

    Args:
        factor (int): The number of elastic oscillations.

    Returns:
        an elastic in ease callable
    """
    return __easing.ease_elastic_in(factor)


def ease_elastic_out(factor: int = 3) -> EasingType:
    r"""Returns an elastic out ease

    .. math::

        \text{ease\_back}(t) = 2^{10(t - 1)}\text{sin}(t\pi/2\times(1 + 4\times\text{factor)})

    Args:
        factor (int): The number of elastic oscillations.

    Returns:
        an elastic out ease callable
    """
    return __easing.ease_elastic_out(factor)


def ease_bounce(bounces: int = 4, damping: float = 0.4) -> EasingType:
    r"""Returns a bounce ease

    Args:
        bounces (int): The number of bounces in the easing.
        damping (float): The energy retained during each bounce.

    Returns:
        a bounce in-out ease callable

    Raises:
        ValueError: damping must be ≥ 0, bounces must be ≥ 1
    """
    return __easing.ease_bounce(bounces, damping)


def ease_bounce_in(bounces: int = 4, damping: float = 0.4) -> EasingType:
    r"""Returns a bounce in ease

    Args:
        bounces (int): The number of bounces in the easing.
        damping (float): The energy retained during each bounce.

    Returns:
        a bounce in ease callable

    Raises:
        ValueError: damping must be ≥ 0, bounces must be ≥ 1
    """
    return __easing.ease_bounce_in(bounces, damping)


def ease_bounce_out(bounces: int = 4, damping: float = 0.4) -> EasingType:
    r"""Returns a bounce out ease

    Args:
        bounces (int): The number of bounces in the easing.
        damping (float): The energy retained during each bounce.

    Returns:
        a bounce out ease callable

    Raises:
        ValueError: damping must be ≥ 0, bounces must be ≥ 1
    """
    return __easing.ease_bounce_out(bounces, damping)


def combine_easing(easing1: EasingType, easing2: EasingType) -> callable:
    """Combines 2 easing functions together.
    
    The 1st easing function is used for t values below 0.5 and the 2nd easing function is used for t values above 0.5

    Args:
        easing1: the 1st easing function for t < 0.5
        easing2: the 2nd easing function for t ≥ 0.5

    Raises:
        TypeError: easing1 and easing2 must be easing functions

    Examples:

        .. code-block:: python

            >>> ease1 = gp.ease_elastic_in()
            >>> ease2 = gp.ease_bounce_out()
            >>> new_ease = combine_easing(ease1, ease2)
    """
    if callable(easing1) and callable(easing2):
        def closure(t: float) -> float:
            """
            Args:
                t: value to use to calculate the easing
            """
            if t < 0.5:
                return easing1(2 * t) / 2
            else:
                return (1 + easing2(2 * t - 1)) / 2

        return closure

    raise TypeError(f"Easing functions expected, got {type(easing1)} and {type(easing2)}")

File: goopylib/src/goopylib/Math/Easing.cpp
#include "gp.h"

#define EASE_IN(func, ...) func(t, ##__VA_ARGS__)
#define EASE_OUT(func, ...) 1 - func(1 - t, ##__VA_ARGS__)
#define EASE_INOUT(func, ...) t < 0.5 ? func(2 * t, ##__VA_ARGS__) / 2 : 1 - func(2 * (1 - t), ##__VA_ARGS__) / 2

namespace {
    void processEaseBounce(int bounces, float *damping, float *scale, std::vector<float> &zeroes,
                           std::vector<float> &derivatives) {
        *damping = -sqrt(*damping);
        zeroes.push_back(1);
        derivatives.push_back(-2);

        zeroes.push_back(-2 * *damping + 1);
        derivatives.push_back(2 * (zeroes[0] - zeroes[1]) + *damping * derivatives[0]);

        for (int i = 2; i < bounces; i++) {
            zeroes.push_back(*damping * (zeroes[i - 2] - zeroes[i - 1]) + zeroes[i - 1]);
            derivatives.push_back(2 * (zeroes[i - 1] - zeroes[i]) + *damping * derivatives[i - 1]);
        }

        *scale = zeroes[bounces - 1];
    }
}

namespace gp {
    std::function<float(float)> easeLinear() {
        return [](float t) {
            return t;
        };
    }

    float inline easePoly(float t, float power) {
        return pow(t, power);
    }

    std::function<float(float)> easePolyIn(float power) {
        return [power](float t) {
            return EASE_IN(easePoly, power);
        };
    }

    std::function<float(float)> easePolyOut(float power) {
        return [power](float t) {
            return EASE_OUT(easePoly, power);
        };
    }

    std::function<float(float)> easePolyInOut(float power) {
        return [power](float t) {
            return EASE_INOUT(easePoly, power);
        };
    }

    float inline easeQuad(float t) {
        return t * t;
    }

    std::function<float(float)> easeQuadIn() {
        return [](float t) {
            return EASE_IN(easeQuad);
        };
    }

    std::function<float(float)> easeQuadOut() {
        return [](float t) {
            return EASE_OUT(easeQuad);
        };
    }

    std::function<float(float)> easeQuadInOut() {
        return [](float t) {
            return EASE_INOUT(easeQuad);
        };
    }

    float inline easeCubic(float t) {
        return t * t * t;
    }

    std::function<float(float)> easeCubicIn() {
        return [](float t) {
            return EASE_IN(easeCubic);
        };
    }

    std::function<float(float)> easeCubicOut() {
        return [](float t) {
            return EASE_OUT(easeCubic);
        };
    }

    std::function<float(float)> easeCubicInOut() {
        return [](float t) {
            return EASE_INOUT(easeCubic);
        };
    }

    float inline easeQuart(float t) {
        return t * t * t * t;
    }

    std::function<float(float)> easeQuartIn() {
        return [](float t) {
            return EASE_IN(easeQuart);
        };
    }

    std::function<float(float)> easeQuartOut() {
        return [](float t) {
            return EASE_OUT(easeQuart);
        };
    }

    std::function<float(float)> easeQuartInOut() {
        return [](float t) {
            return EASE_INOUT(easeQuart);
        };
    }

    float inline easeQuint(float t) {
        return t * t * t * t * t;
    }

    std::function<float(float)> easeQuintIn() {
        return [](float t) {
            return EASE_IN(easeQuint);
        };
    }

    std::function<float(float)> easeQuintOut() {
        return [](float t) {
            return EASE_OUT(easeQuint);
        };
    }

    std::function<float(float)> easeQuintInOut() {
        return [](float t) {
            return EASE_INOUT(easeQuint);
        };
    }

    float inline easeCircle(float t) {
        return 1 - sqrt(1 - (float) pow(t, 2));
    }

    std::function<float(float)> easeCircleIn() {
        return [](float t) {
            return EASE_IN(easeCircle);
        };
    }

    std::function<float(float)> easeCircleOut() {
        return [](float t) {
            return EASE_OUT(easeCircle);
        };
    }

    std::function<float(float)> easeCircleInOut() {
        return [](float t) {
            return EASE_INOUT(easeCircle);
        };
    }

    float inline easeSin(float t) {
        return 1 - (float) cos(t * 1.57079632679);
    }

    std::function<float(float)> easeSinIn() {
        return [](float t) {
            return EASE_IN(easeSin);
        };
    }

    std::function<float(float)> easeSinOut() {
        return [](float t) {
            return EASE_OUT(easeSin);
        };
    }

    std::function<float(float)> easeSinInOut() {
        return [](float t) {
            return EASE_INOUT(easeSin);
        };
    }

    float inline easeExp(float t, float base) {
        return pow(base, 10 * t - 10);
    }

    std::function<float(float)> easeExpIn(float base) {
        return [base](float t) {
            return EASE_IN(easeExp, base);
        };
    }

    std::function<float(float)> easeExpOut(float base) {
        return [base](float t) {
            return EASE_OUT(easeExp, base);
        };
    }

    std::function<float(float)> easeExpInOut(float base) {
        return [base](float t) {
            return EASE_INOUT(easeExp, base);
        };
    }

    float inline easeBack(float t, float factor) {
        return ((factor + 1) * t * t * t - factor * t * t);
    }

    std::function<float(float)> easeBackIn(float factor) {
        return [factor](float t) {
            return EASE_IN(easeBack, factor);
        };
    }

    std::function<float(float)> easeBackOut(float factor) {
        return [factor](float t) {
            return EASE_OUT(easeBack, factor);
        };
    }

    std::function<float(float)> easeBackInOut(float factor) {
        return [factor](float t) {
            return EASE_INOUT(easeBack, factor);
        };
    }

    float inline easeElastic(float t, int factor) {
        return (float) pow(2, 10 * (t - 1)) * (float) std::sin(1.5707963267 * t * (1 + 4 * factor));
    }

    std::function<float(float)> easeElasticIn(int factor) {
        return [factor](float t) {
            return EASE_IN(easeElastic, factor);
        };
    }

    std::function<float(float)> easeElasticOut(int factor) {
        return [factor](float t) {
            return EASE_OUT(easeElastic, factor);
        };
    }

    std::function<float(float)> easeElasticInOut(int factor) {
        return [factor](float t) {
            return EASE_INOUT(easeElastic, factor);
        };
    }

    float inline easeBounce(float t, float damping, int bounces, float scale, const std::vector<float> &zeroes,
                            const std::vector<float> &derivatives) {
        t = scale * (1 - t);
        for (int i = 0; i < bounces; i++) {
            if (zeroes[i] >= t) {
                if (i == 0) {
                    return -t * t + 1;
                }
                else {
                    t -= zeroes[i - 1];
                    return -t * t + damping * t * derivatives[i - 1];
                }
            }
        }
        return 1;
    }

    std::function<float(float)> easeBounceIn(int bounces, float damping) {
        std::vector<float> zeroes, derivatives;
        float scale;
        processEaseBounce(bounces, &damping, &scale, zeroes, derivatives);

        return [damping, bounces, scale, zeroes, derivatives](float t) {
            return EASE_IN(easeBounce, damping, bounces, scale, zeroes, derivatives);
        };
    }

    std::function<float(float)> easeBounceOut(int bounces, float damping) {
        std::vector<float> zeroes, derivatives;
        float scale;
        processEaseBounce(bounces, &damping, &scale, zeroes, derivatives);

        return [damping, bounces, scale, zeroes, derivatives](float t) {
            return EASE_OUT(easeBounce, damping, bounces, scale, zeroes, derivatives);
        };
    }

    std::function<float(float)> easeBounceInOut(int bounces, float damping) {
        std::vector<float> zeroes, derivatives;
        float scale;
        processEaseBounce(bounces, &damping, &scale, zeroes, derivatives);

        return [damping, bounces, scale, zeroes, derivatives](float t) {
            return EASE_INOUT(easeBounce, damping, bounces, scale, zeroes, derivatives);
        };
    }
}
